### PR TITLE
Upgrade mono to support GitLab-backed repositories.

### PR DESCRIPTION
**Pull request by stristr**
OK to start testing in conjunction with [Inkstone #23](https://github.com/HaikuTeam/inkstone/pull/23). This will be a hefty omnibus review.

Changes in this PR:

- [x] [Inkstone 3372a55](https://github.com/HaikuTeam/inkstone/pull/23/commits/3372a5582870f8bb92ab741ee6e0dc3dd6c3f279) attaches a `RepositoryUrl` field to all *authenticated* calls to retrieve a list of projects if the project is GitLab-backed which carries the necessary credentials to clone using HTTP auth. In Plumbing, we capture it in and bubble it up to all the places that need to use it.
 - [x] When we detect a GitLab-backed project, use a new code path that does not expect credentials to be separately received and managed except when managing remotes.
 - [x] In this new code path, provide upgrade-in-place during remote upsertion for projects that started on CodeCommit but have been migrated in Inkstone.
 - [x] Overall speed up project init by removing some bloat and parallelizing some work that was parallelize-able.
 - [x] Wipe out the step where we rsync `@haiku/player` into the project folder. It hasn't been functional in production for a while, and shouldn't be needed. We should revisit with live design, but definitely benefit from the savings here and now.
 - [x] Upgrade `haiku clone` in `@haiku/cli` to work with GitLab remotes.
 - [x] Fix some collab features:
   - [x] If project doesn't exist locally, clone it.
   - [x] If project has upstream changes, merge them (using an aggressive `--ours` strategy if necessary) while opening project. (This doesn't seem to slow things down very much at all, unless there are merge conflicts—which should be resolved, else pushing won't work as desired).
 - [x] Fix some edge bugs that were tangentially related:
   - [x] Deleting a project, then creating a project with the same name now works correctly. (We archive the old project directory in `~/.haiku` because #neverdelete.)

[Plumbing tests are currently down](https://app.asana.com/0/506410768732346/514402868389861/f); I plan to get in there and fix them before this goes out. Meanwhile, these are the test cases we should be slamming on:
 - The entire app works for project CRUD deployed against current production Inkstone (no GitLab work):
   - [ ] Create new project, make Haiku, publish, get GIF via PhantomJS
   - [ ] Open existing project, make changes, publish, get GIF via PhantomJS
   - [ ] Delete project
 - The entire app works for project CRUD deployed against Inkstone branch `feat-gitlab` for an existing user:
   - [ ] Create new (GitLab-backed) project, make Haiku, publish, get GIF via Puppeteer
   - [ ] Open existing project that has not been migrated, make changes, publish, get GIF via PhantomJS
   - [ ] Migrate CodeCommit project while it's open in the app on Inkstone (`/gocode/src/inkstone/inkstone scripts migration-organization-gitlab --orgID <N>`), back out, reopen the project. Confirm remotes point to GitLab in project folder and publish goes GitLab post-receive hook and GIF is via Puppeteer.
   - [ ] Open CodeCommit-backed CheckTutorial project, make changes, publish, get GIF via PhantomJS
   - [ ] Delete CodeCommit-backed project
   - [ ] Delete GitLab-backed project
- The entire app works for project CRUD for a brand new user.
   - [ ] Create new (GitLab-backed) project, make Haiku, publish, get GIF
   - [ ] Open GitLab-backed CheckTutorial project, make changes, publish, get GIF
   - [ ] Delete GitLab-backed project

### CODE REVIEW

**Comment by stristr** [comment]
The first set of QA criteria (everything works without updating Inkstone) should be ready for someone's careful review. Essentially, just check out this branch, and confirm everything works:
 - Create new project, make Haiku, publish, get GIF
 - Open existing project, publish, get GIF
 - Delete project
 - Bonus: in the same session, recreate a project of the same name. It should come up blank with a backup in your `~/.haiku`.

**Comment by stristr** [comment]
If the above criteria pass, I would like to merge this to unblock GitLab launch.

**Comment by tsortwehttam** [comment]
FWIW, I saw this error when deleting a project:

```
2018-01-17T23:34:53.525Z - error:  TypeError: File.destroyWhere is not a function
    at Function.expelOne (/Users/matthew/Code/HaikuTeam/mono/packages/haiku-serialization/src/bll/File.js:846:15)
    at /Users/matthew/Code/HaikuTeam/mono/packages/haiku-plumbing/lib/Master.js:516:35
    at /Users/matthew/Code/HaikuTeam/mono/packages/haiku-plumbing/lib/MasterGitProject.js:1193:31
    at /Users/matthew/Code/HaikuTeam/mono/packages/haiku-plumbing/lib/MasterGitProject.js:1181:16
    at /Users/matthew/Code/HaikuTeam/mono/packages/haiku-plumbing/lib/MasterGitProject.js:1156:18
    at done (/Users/matthew/Code/HaikuTeam/mono/packages/haiku-plumbing/lib/Git.js:154:14)
    at /Users/matthew/Code/HaikuTeam/mono/packages/haiku-plumbing/lib/Git.js:158:23
    at /Users/matthew/Code/HaikuTeam/mono/packages/haiku-plumbing/lib/Git.js:132:21
    at tryCallOne (/Users/matthew/Code/HaikuTeam/mono/node_modules/nodegit-promise/lib/core.js:37:12)
    at /Users/matthew/Code/HaikuTeam/mono/node_modules/nodegit-promise/lib/core.js:103:15
    at flush (/Users/matthew/Code/HaikuTeam/mono/node_modules/asap/raw.js:50:29)
    at _combinedTickCallback (internal/process/next_tick.js:73:7)
    at process._tickDomainCallback (internal/process/next_tick.js:128:9)
```

But I only saw it occur once across 4 deletions.

**Comment by stristr** [comment]
This seems to be a tangential side effect of a lingering file watcher but points out an actual bug…I added code that archives the project with `fse.move`, but this is triggered by `Master#handleFileRemove` => `File#expelOne` => `File.destroyWhere`, which is indeed not a function (likely removed in a distant refactor, so good that you caught this). So…why is the watcher still active when you're back on the project dashboard in a place where you can press delete?

File: packages/haiku-plumbing/src/Plumbing.js [hunk 1]
```diff
@@ -799,18 +783,34 @@ export default class Plumbing extends StateObject {
...
     logger.info('[plumbing] deleting project', name)
     const authToken = sdkClient.config.getAuthToken()
     return inkstone.project.deleteByName(authToken, name, (deleteErr) => {
       if (deleteErr) {
         this.sentryError('deleteProject', deleteErr)
         if (cb) return cb(deleteErr)
       }
+      if (fse.existsSync(path)) {
+        // Delete the project locally, but in a recoverable state.
+        let archivePath = `${path}.bak`
```
Comment (tsortwehttam) [inline packages/haiku-plumbing/src/Plumbing.js:796]: Neat.

**Comment by tsortwehttam** [comment]
Code LGTM.

Also did the review tasks you outlined and those also LGTM.

**Pull request merged**